app/src/main/java/ch/epfl/skysync/viewmodel/CalendarViewModel.kt

```diff
@@ -0,0 +1,143 @@
+package ch.epfl.skysync.viewmodel
+
+import androidx.compose.runtime.Composable
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.ViewModelProvider
+import androidx.lifecycle.viewModelScope
+import androidx.lifecycle.viewmodel.compose.viewModel
+import ch.epfl.skysync.database.ParallelOperationsEndCallback
+import ch.epfl.skysync.database.tables.AvailabilityTable
+import ch.epfl.skysync.database.tables.UserTable
+import ch.epfl.skysync.models.calendar.AvailabilityCalendar
+import ch.epfl.skysync.models.calendar.CalendarDifferenceType
+import ch.epfl.skysync.models.calendar.FlightGroupCalendar
+import ch.epfl.skysync.models.user.User
+import ch.epfl.skysync.util.WhileUiSubscribed
+import com.google.firebase.auth.FirebaseUser
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.combine
+import kotlinx.coroutines.flow.stateIn
+
+data class CalendarUiState(
+    val user: User? = null,
+    val availabilityCalendar: AvailabilityCalendar = AvailabilityCalendar(),
+    val flightGroupCalendar: FlightGroupCalendar = FlightGroupCalendar(),
+    val isLoading: Boolean = false,
+)
+
+/**
+ * ViewModel for the Calendar screen
+ *
+ * @param firebaseUser The Firebase user
+ * @param userTable The user table
+ * @param availabilityTable The availability table
+ */
+class CalendarViewModel(
+    firebaseUser: FirebaseUser,
+    private val userTable: UserTable,
+    private val availabilityTable: AvailabilityTable,
+) : ViewModel() {
+  companion object {
+    /** Creates a view model by accepting the firebase user as an argument */
+    @Composable
+    fun createViewModel(
+        firebaseUser: FirebaseUser,
+        userTable: UserTable,
+        availabilityTable: AvailabilityTable,
+    ): CalendarViewModel {
+      return viewModel<CalendarViewModel>(
+          factory =
+              object : ViewModelProvider.Factory {
+                override fun <T : ViewModel> create(modelClass: Class<T>): T {
+                  return CalendarViewModel(firebaseUser, userTable, availabilityTable) as T
+                }
+              })
+    }
+  }
+
+  private val user: MutableStateFlow<User?> = MutableStateFlow(null)
+  private val loadingCounter = MutableStateFlow(0)
+
+  private val uid: String
+  private var originalAvailabilityCalendar = AvailabilityCalendar()
+
+  val uiState: StateFlow<CalendarUiState> =
+      combine(user, loadingCounter) { user, loadingCounter ->
+            CalendarUiState(
+                user,
+                user?.availabilities ?: AvailabilityCalendar(),
+                user?.assignedFlights ?: FlightGroupCalendar(),
+                loadingCounter > 0)
+          }
+          .stateIn(
+              scope = viewModelScope,
+              started = WhileUiSubscribed,
+              initialValue = CalendarUiState(isLoading = true))
+
+  init {
+    uid = firebaseUser.uid
```

Comment:
uid could be passed directly as argument since firebaseUser is only used as a container for the uid?